bouncer/src/repo/core/model/bson/repo_bson_factory.cpp

```diff
@@ -25,6 +25,203 @@
 
 using namespace repo::core::model;
 
+/**
+* Specialisation of the BSON Builder dedicated to building Meshes and
+* Supermeshes. This is for internal use by the RepoBSONFactory and is
+* intended to build immutable nodes via one of the makeMeshNode
+* overloads.
+*/
+class RepoMeshBSONBuilder : public RepoBSONBuilder
+{
+	using BinMapping = std::unordered_map<std::string, std::pair<std::string, std::vector<uint8_t>>>;
+public:
+	std::string uniqueID;
+
+	BinMapping& mapping()
+	{
+		return binMapping;
+	}
+
+	size_t& bytesize() {
+		return _byteSize;
+	}
+
+	/*
+	* TODO: because mongo has a stupid internal limit of 64MB, we can't store everything in a BSON
+	* There are 2 options
+	* 1. store binaries in memory outside of the bson and put it into GRIDFS at the point of commit
+	* 2. leave mongo's bson, use our own/exteral library that doesn't have this limit and database handle this at the point of commit
+	* below uses option 1, but ideally we should be doing option 2.
+	*/
+
+	template<typename T>
+	void appendLargeArray(std::string name, std::vector<T> data)
+	{
+		uint64_t byteCount = data.size() * sizeof(data[0]);
+
+		if (byteCount + _byteSize >= REPO_BSON_MAX_BYTE_SIZE) //inclusion of this binary exceeds the maximum, store separately
+		{
+			std::string bName = uniqueID + "_" + name;
+			binMapping[name] =
+				std::pair<std::string, std::vector<uint8_t>>(bName, std::vector<uint8_t>());
+			binMapping[name].second.resize(byteCount); //uint8_t will ensure it is a byte addrressing
+			memcpy(binMapping[name].second.data(), &data[0], byteCount);
+			_byteSize += sizeof(bName);
+		}
+		else
+		{
+			appendBinary(
+				name,
+				&data[0],
+				data.size() * sizeof(data[0]));
+			_byteSize += byteCount;
+		}
+	}
+
+	void appendDefaults(
```

Comment:
from this part on, this is getting slightly weird because it's very much what the bson_factory was responsible for

Whereas bson_builder is more designed to be alight interface to building a bson, with understanding of 3drepo types.

for e.g appendDefaults is about appending the default information required for a RepoNode,  hence it lives in RepoBSONFactory as that has nothing to do with the concept of building a RepoBSON - it's a context of a specific type of repoBson (i.e. RepoNode)

So I'm starting to wonder why you feel like you need this new class - is this just to split the logic into multiple functions so you can pick and mix what it needs to do for mesh/supermesh?

I'm thinking it may be better instead to just have a helper function that takes a bsonbuilder and the data to build the common parts?

Reply:
Or call this mesh node builder and remove the inheritance and have a repobsonbuilder as  an instance?

The blurring of roles between the factory and bson builder is what's confusing me a bit, not necessarily the concept of having a helper class here!

Reply:
Hi Carmen, ah I understand...

I have turned `RepoMeshBSONBuilder` into `RepoBSONBinMappingBuilder` and moved it to its own file. I have made it responsible for managing the binary mappings internally, and kept the `appendLargeArray` method, but the other methods have been made private members of `RepoBSONFactory`. 
I have also made sure there is only one set of `appendDefaults` methods, common to the factory, that the mesh building methods use.